### ISSUE TITLE
Improve "no-constant-condition" rule for generators

### ISSUE THREAD

**Issue by Turbo87** [eslint/eslint#8566]
This topic has been discussed already in https://github.com/eslint/eslint/issues/1299 and https://github.com/eslint/eslint/issues/2375 but per https://github.com/eslint/eslint/issues/2375#issuecomment-299883115 I'm opening a new issue for this.

https://github.com/eslint/eslint/issues/2375#issuecomment-255137510 has introduced a proposal for handling this which seems pretty reasonable:

Bad:

```js
while(true) {
    //...
}
```

Bad:

```js
function*() {
    while(true) {
        //...
    }
    yield 'foo';
}
```

Good:

```js
function*() {
    while(true) {
        yield 'foo';
    } 
}
```

~It should be noted that the same should also apply to async functions which have roughly similar semantics. Since both generators and async/await are starting to be adopted a lot more this topic should be revisited.~

/cc @platinumazure

**Comment by platinumazure** [comment]
So basically, my take on it is this:

* The purpose of no-constant-condition is to avoid useless conditionals and infinite loops.
* However, `while(true)` in a generator, as long as at least one code path (or all code paths? not sure) has a `yield` statement in it, is not actually an infinite loop and serves a different purpose, basically implementing an infinite generator or coroutine that does not actually do an infinite loop.
* I assume similar logic holds for `await` within an async function, but I will defer to more knowledgeable folks on that.
* Our last decision on the issue was that generators and async/await had not yet shown themselves to be mainstream and `while(true)` in generators especially had yet to prove itself. Since then, async/await is in ES2017 and `while(true)` in generators is definitely becoming more common.

So I think this merits revisiting, and I will :+1: this enhancement.

**Comment by not-an-aardvark** [comment]
> I assume similar logic holds for await within an async function, but I will defer to more knowledgeable folks on that.

This is incorrect -- when generator functions use `yield`, control flow returns to the code that invoked the generator. However, when an async function uses `await` in an infinite loop, the returned Promise never fulfills.

**Comment by Turbo87** [comment]
@not-an-aardvark that is a good point. I guess we should limit this to just generator for now then and discuss async functions at a later point.

**Comment by Andarist** [comment]
@not-an-aardvark 
how so? could u post a simple snippet with the explanation? I think thats not true and ive just tested it in chrome

```javascript
function delay(ms) {
  const promise = new Promise(resolve => {
    setTimeout(() => resolve(), ms)
  })

  return promise
}

async function backoff(i = 0) {
  while (true) {
    console.log(i++)
    await delay(i * 1000)
    
  }
}

backoff()
```

**Comment by Turbo87** [comment]
@Andarist the promise returned from backoff() will never resolve which  doesn't seem like good API design 🤔

**Comment by Andarist** [comment]
Hm. I guess... although i dont see this as such bad design if somebody really needs an infinite task

**Comment by platinumazure** [comment]
@Andarist My take on that is, just like for infinite loops, `no-constant-condition` can't know when you mean to have an infinite task or not so your best bet is to use `// eslint-disable-line` comments on those cases.

Infinite generators, on the other hand, are a completely different animal altogether and I think we should consider enhancing `no-constant-condition` to allow an option to exclude infinite loops in generator functions (as long as they have `yield` statements in them).

**Comment by benjamingr** [comment]
`while` loops in `async` functions can serve a purpose:

```js
// background worker
(async () => {
  while(true) {
    await delay(15 * 1000);
    updateStockPrices();
  }
})();
```

That is, global "background work". That said, the above code can be rewritten as:

```js
setInterval(updateStockPrices, 15000);
```

On the other hand - async generators should definitely be supported (functions that are both async and generators).

**Comment by not-an-aardvark** [comment]
@eslint/eslint-team We have 3 :+1:s for this proposal, so we just need a team member to champion it so we can accept it.

**Comment by platinumazure** [comment]
I'll champion if we're okay with restricting this to generators for now (and if @Turbo87 is willing to modify the original post and issue title accordingly to avoid confusion).

That means we would need one more :+1: from the team since I'm currently one of the three.

**Comment by Turbo87** [comment]
@platinumazure I've updated the title and post. Thanks for the reminder :)

**Comment by VictorHom** [comment]
If possible, can I take on this issue?

**Comment by kaicataldo** [comment]
@VictorHom Please feel free - and thank you!
